### ISSUE TITLE
DOM: `Element#matches()` incorrectly narrows types

### ISSUE THREAD

**Issue by Derugon** [microsoft/TypeScript#63497]
### 🔎 Search Terms

element matches narrow never

### 🕗 Version & Regression Information

- This changed between versions 5.9.3 and 6.0.3

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=6.0.3#code/CYUwxgNghgTiAEEQBd4iQWxAO2QLngAkAVAWQBkBRTHZAbgCgBCdELXAOgymTAAsQAZwAU8AOTBkY+AEp4AMnloanbrwEjxwYNJlA

### 💻 Code

```ts
declare let element: HTMLElement;
!element.matches( 'dt' ) && element.matches( 'dd' )
```

### 🙁 Actual behavior

`element.matches( 'dd' )` throws a `Property 'matches' does not exist on type 'never'` compilation error.

### 🙂 Expected behavior

`element.matches( 'dd' )` does not throw any error.

### Additional information about the issue

`element.matches( 'dt' )` uses the following overload:
```ts
matches<K extends keyof HTMLElementTagNameMap>(selectors: K): this is HTMLElementTagNameMap[K];
```

The return type of `element.matches( 'dt' )` is then `this is HTMLElementTagNameMap['dt']` that maps to `this is HTMLElement`.

Since `this` is `HTMLElement`:
- When false, the type of `element` is `HTMLElement` (not narrowed further).
- When false, the type of `element` is `never`.

**Comment by MartinJohns** [comment]
This is due to this change: https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/2147

A workaround would be to use the `string` overload: `element.matches('dt' as string)`

**Comment by RyanCavanaugh** [comment]
🤖 Thank you for your issue! I've done some analysis to help get you started. This response is automatically generated; feel free to 👍 or 👎 this comment according to its usefulness.

## Similar Issues

Here are the most similar issues I found
- (67%) [microsoft/typescript#5813](https://redirect.github.com/microsoft/typescript/issues/5813): Property 'matches' does not exist on type 'Element'
- (67%) [microsoft/typescript#49454](https://redirect.github.com/microsoft/typescript/issues/49454): element\.matches\(\) can be array of strings

If your issue is a duplicate of one of these, feel free to close this issue. Otherwise, no action is needed.

## Possibly Working as Intended

It looks like this behavior might be intentional. Here's what a computer had to say about it.

**Comment by nmain** [comment]
```
matches<K extends keyof HTMLElementTagNameMap>(selectors: K): this is HTMLElementTagNameMap[K];
```

Because typeguards are two-sided, this would only be correct if each member of `HTMLElementTagNameMap` was a structurally distinct subclass, but that's not the case.

**Comment by RyanCavanaugh** [comment]
The provided repro is a little weird since there's no reason to write code that way, but you could imagine e.g.
```ts
declare let element: HTMLElement;
if (element.matches('dt')) {
    // ...
} else {
    // uh oh
    const s: string = element;
}
```

**Comment by voxpelli** [comment]
> This is due to this change: https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/2147
> 
> A workaround would be to use the `string` overload: `element.matches('dt' as string)`

~~That PR didn't ship in TS 6.0, it shipped earlier? So it's either due to something that changed in TS 6.0 or it's been there since before 6.0?~~

It did apparently first ship in TS 6.0 I was wrong.

**Comment by voxpelli** [comment]
Proper fix for it (on my phone and about to sleep so can't make PR now but should work):

```ts
type SpecificHTMLElementTagNameMap = {
    [K in keyof HTMLElementTagNameMap as HTMLElement extends HTMLElementTagNameMap[K] ? never : K]: HTMLElementTagNameMap[K];
};
type SpecificSVGElementTagNameMap = {
    [K in keyof SVGElementTagNameMap as SVGElement extends SVGElementTagNameMap[K] ? never : K]: SVGElementTagNameMap[K];
};

matches<K extends keyof SpecificHTMLElementTagNameMap>(selectors: K): this is SpecificHTMLElementTagNameMap[K];
matches<K extends keyof SpecificSVGElementTagNameMap>(selectors: K): this is SpecificSVGElementTagNameMap[K];
matches(selectors: string): boolean;

```

MathML excluded intentionally, it can never work with narrowing as it is now

**Comment by Derugon** [comment]
> The provided repro is a little weird since there's no reason to write code that way

Yes, I'm sorry, the code above was just a small chunk highlighting the issue. The actual use case is a sequence of conditionals, choosing which method to call depending on selector matches, either with a single element name or multiple checks, for example:

```ts
declare let element: HTMLElement;
if ( element.matches( 'th[scope=row], th[scope=rowgroup], td.cell-base' ) ) {
	// ok
} else if ( element.matches( 'dt' ) ) {
	// ok
} else if ( element.matches( 'h2, h3, h4, h5, h6, .headline' ) ) {
	// Error: Property 'matches' does not exist on type 'never'
}
```

**Comment by Derugon** [comment]
> Proper fix for it (on my phone and about to sleep so can't make PR now but should work):

The issue applies when `element.matches()` on an `element: X` returns `this is X`, and not `this is Y` (with `Y != X`). For example, this fix would still allow narrowing `HTMLTableCellElement` to itself:

[*(playground link)*](https://www.typescriptlang.org/play/?ts=6.0.3#code/CYUwxgNghgTiAEEQBd4iQWxAO2QLngAkAVAWQBlioAjJAYXQgFFMdkBuAKAEsAzeABToQWXADoMUZGAAWIAM4CA5MhlKAlOvgBvTgEgA9AfhjTnAL5oI8hLr1gA9tnmpuweAF4rItmLddzIA)
```ts
declare let element: HTMLTableCellElement;
if (element.matches('th')) {
	// ...
} else {
	const id = element.id; // Error: Property 'id' does not exist on type 'never'
}
```

Below are a few test cases checking for narrowing and this issue (this should compile):

[*(playground link)*](https://www.typescriptlang.org/play/?ts=6.0.3#code/CYUwxgNghgTiAEYD2A7AzgF3iCIC2IKGAXPAKK4FEDcAUKJLAsulgBYZ4QX6EnwAJACoBZADI8qGOg2hxEqTPA5chUAEa4Awjm6U+pYeLWaQOiHt41atAPS34AOQCMpSX0RQU8dQhSwYJAB3EGB4DCRBUQl9IgA6WgBLADN4AAocKww4vCgMMDYQNDSAcmAMEoBKSvgAb1oASBYlRLRHAODQw2j3IngAXmxY6VoAXzsHMlco8V6sNDYkAFcIMJQkLF94fxhAkLCImZishJT0gEIVSykcvIKi0vKqmvqmxSxWxw3HEAA3EBgpAwAE8AA4gJCpK5zAbbEBBI5zNKVOjjexOABM3Vmw083i2Oz2oXCkSMYhM2l0c1OqTS0OGt3yhWKJQwwGedUazQ+bQ6+2x5I0lIsMMG9KyqIm5CxRwpZipuIWy1W2w2Pj8fOJhzJcvM1z4NIuV11CpOuSZD1ZbA5r258E+3z+AKBYIhUM4EBNItxgxQ8NlQvl3qyyNRQA)
```ts
declare const element: Element;
declare const htmlElement: HTMLElement;
declare const htmlTableCellElement: HTMLTableCellElement;

// N1: Element can be narrowed to HTMLElement.
if (element.matches('dt')) {
	const isNarrowed: HTMLElement = element;
}
// E1: HTMLElement should not be narrowed to HTMLElement.
if (!htmlElement.matches('dt')) {
	const isNotNever: typeof htmlElement = new HTMLElement();
}
// N2: HTMLElement can be narrowed to HTMLTableCellElement.
if (htmlElement.matches('td')) {
	const isNarrowed: HTMLTableCellElement = htmlElement;
}
// E2: HTMLTableCellElement should not be narrowed to HTMLTableCellElement.
if (!htmlTableCellElement.matches('th')) {
	const isNotNever: typeof htmlTableCellElement = new HTMLTableCellElement();
}
```

**Comment by voxpelli** [comment]
When are you encountering these issues in practice?

Eg here's another example that seems to work fine: https://www.typescriptlang.org/play/?ts=6.0.3#code/MYewdgzgLgBApgGzgWwEwwLwwCYmAV2TjCgDoBHfOAJwE8BlROYKEagCgHIptOBKANwAoAJYAzGOyZoA-KWQBDKMAAWcCFx78+MAN5CAkKEiwxIEAC4YACQAqAWQAythQCMkAYUQIAokiIkmPD+qMIAvsEQcDDiktKocorKahrcKtp6hsbQMGaWNg7Obp7efijEsFjx4ZHRsVIhiUqq6lzYIgBuGfpG4Dl5VnZOACKdZQGVwSihQhGIUZm9Jrnmg4XjFUHVs0A

**Comment by MartinJohns** [comment]
> Eg here's another example that seems to work fine:

It only works fine because you're using the null-conditional operator. Have it non-null or explicitly check for `null` and it will fail again.

**Comment by voxpelli** [comment]
Going back to the intent of the change:

To be able to apply similar deduction of HTML element type for an existing element as `querySelectorAll('div')` / `querySelector('div')`, providing an option to `instanceof HTMLDivElement` and such.

But it seems like that could then only ever work for tags that have a 1:1 relationship to their class then.

**Comment by Derugon** [comment]
A solution could be using `this is …` on the subset of tags whose interface extends `this` and are not exactly `this`, such as:

```ts
/**
 * From T, pick a set of properties whose values extend U
 */
type PickExtends<T, U> = {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

/**
 * From T, pick a set of properties whose values extend U, excluding U itself
 */
type PickExtendsStrict<T, U> = {
    [K in keyof PickExtends<T, U> as U extends T[K] ? never : K]: T[K];
};

declare interface Element {
	matches<K extends keyof PickExtendsStrict<HTMLElementTagNameMap, this>>(selectors: K): this is HTMLElementTagNameMap[K];
	matches<K extends keyof PickExtendsStrict<SVGElementTagNameMap, this>>(selectors: K): this is SVGElementTagNameMap[K];
	matches<K extends keyof PickExtendsStrict<MathMLElementTagNameMap, this>>(selectors: K): this is MathMLElementTagNameMap[K];
	matches(selectors: string): boolean;
}
```

However, unlike the merged PR, this would only provide autocompletion for tag names that use interfaces (e.g. `htmlElement.matches(…)` will not provide autocompletion for `'dt'`). A solution could be adding another overload:

```ts
declare interface Element {
	matches<K extends keyof PickExtendsStrict<HTMLElementTagNameMap, this>>(selectors: K): this is HTMLElementTagNameMap[K];
	matches<K extends keyof PickExtendsStrict<SVGElementTagNameMap, this>>(selectors: K): this is SVGElementTagNameMap[K];
	matches<K extends keyof PickExtendsStrict<MathMLElementTagNameMap, this>>(selectors: K): this is MathMLElementTagNameMap[K];
	matches<K extends keyof PickExtends<HTMLElementTagNameMap | SVGElementTagNameMap | MathMLElementTagNameMap, this>>(selectors: K): boolean;
	matches(selectors: string): boolean;
}
```

**Comment by LeonxLJX** [comment]
I'd like to take this one — I'll follow up with a PR. (claiming via @LeonxLJX)

**Comment by LeonxLJX** [comment]
Hi! I'd like to fix `Element#matches()` narrowing. Plan: reproduce, correct the narrowing, add a test. May I be assigned?
